storage/src/test/java/org/apache/kafka/storage/internals/log/LocalLogTest.java

```diff
@@ -0,0 +1,785 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.storage.internals.log;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.compress.Compression;
+import org.apache.kafka.common.config.TopicConfig;
+import org.apache.kafka.common.errors.KafkaStorageException;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.SimpleRecord;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.server.util.MockTime;
+import org.apache.kafka.test.TestUtils;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.channels.ClosedChannelException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+
+class LocalLogTest {
+
+    private static final MockTime MOCK_TIME = new MockTime();
+
+    private final File tmpDir = TestUtils.tempDirectory();
+    private final File logDir = TestUtils.randomPartitionLogDir(tmpDir);
+    private final TopicPartition topicPartition = new TopicPartition("test_topic", 1);
+    private final LogDirFailureChannel logDirFailureChannel = new LogDirFailureChannel(10);
+
+    private LocalLog log;
+
+    @BeforeEach
+    public void setUp() throws IOException {
+        log = createLocalLogWithActiveSegment(logDir, new LogConfig(new Properties()));
+    }
+
+    @AfterEach
+    public void tearDown() throws IOException {
+        try {
+            log.close();
+        } catch (KafkaStorageException kse) {
+            // ignore
+        }
+        Utils.delete(tmpDir);
+        Utils.delete(logDir);
+    }
+
+    static class KeyValue {
```

Comment:
this can be simplified by `record` class - 
```java
    record KeyValue(String key, String value) {
        SimpleRecord toRecord(long timestamp) {
            return new SimpleRecord(timestamp, key.getBytes(), value.getBytes());
        }

        SimpleRecord toRecord() {
            return new SimpleRecord(MOCK_TIME.milliseconds(), key.getBytes(), value.getBytes());
        }

        static KeyValue fromRecord(Record record) {
            String key = record.hasKey()
                ? StandardCharsets.UTF_8.decode(record.key()).toString()
                : "";
            String value = record.hasValue()
                ? StandardCharsets.UTF_8.decode(record.value()).toString()
                : "";
            return new KeyValue(key, value);
        }
    }
```

Reply:
Right, I guess we can start using `record` classes, updated. Thanks